### ISSUE TITLE
Bad performance and memory usage on DX12

### ISSUE THREAD

**Issue by DasEtwas** [gfx-rs/wgpu#1158]
**Description**
The traced application runs poorly on a GTX 970 but it performs fine using the vulkan abckend.

**Repro steps**
Play the trace.

**Expected vs observed behavior**
Attached is a trace of a wgpu application running via DirectX 12 on a GTX 970.
Frame times should be lower than 0.02s. Application memory usage should be nearly static, levelling off at ~300-400MB, with only new planet chunk textures being allocated causing a deviation of at most 50 MB in CPU and GPU memory.
When not tracing, the application actually hangs after rendering two to three frames.

Memory usage is very high, with a constant rise of about 100MB per second or more in CPU memory.
Even when not tracing wgpu, frame times are about ~0.3s compared to ~0.008s via Vulkan.

**Platform**
`wgpu = { git = "https://github.com/gfx-rs/wgpu-rs", rev = "baf6c433bb8daa230c350c777e5f26fc95382495", features = ["trace"] }`

profile:
```
opt-level = 2
debug = true
debug-assertions = false
overflow-checks = true
lto = false
incremental = true
codegen-units = 64
````

trace: https://www.dropbox.com/s/bp1e9btyf9beeaw/wgpu_trace.zip

**Comment by baptistemanson** [comment]
Hi @DasEtwas , 
I ran your trace with wgpu c788 (he one used by wgpu-rs baf6c), and it crashes after opening a window and painting a black rectangle on the right hand side of the window:
```
thread 'main' panicked at 'Index 19 is already occupied', C:\...\wgpu-core\src\hub.rs:168:18
```
Maybe related to your previous finding https://github.com/gfx-rs/wgpu/issues/1146 ?

Hope that helps anyone in diagnosing something.

**Comment by DasEtwas** [comment]
> Hi @DasEtwas , 
> I ran your trace with wgpu c788 (he one used by wgpu-rs baf6c), and it crashes after opening a window and painting a black rectangle on the right hand side of the window:
> ```
> thread 'main' panicked at 'Index 19 is already occupied', C:\...\wgpu-core\src\hub.rs:168:18
> ```
> Maybe related to your previous finding https://github.com/gfx-rs/wgpu/issues/1146 ?
> 
> Hope that helps anyone in diagnosing something.

I encountered the same problem. I think it is an unrelated bug because the application starts off with rendering hundreds of frames for a loading screen which cannot possibly trigger that problem. You can verify this by changing the backend everywhere in the trace.ron to Vulkan, which shows the same behaviour (and it trips an allocator assertion).

**Comment by kvark** [comment]
#1161 fixes the replay issue

**Comment by baptistemanson** [comment]
With the fix by @kvark, I was able to run your trace with no major problem.
I have a 2080 SUPER. On my card, the most time consuming thing I saw was the different calls to the swap context. It was blocking my main thread around 30% of the time.
Here is my superluminal profiling session if someone has more experience than me: https://drive.google.com/file/d/1ZEcsFHpVrVZI9d8P4swyQ2Xhhx_l9fla/view?usp=sharing
